File: backend/app/schemas/recurring.py
from datetime import date, datetime
from typing import Annotated

from pydantic import Field, model_validator

from app.domain.constants import (
    RecurrenceMode,
    RecurringBillingMode,
    RecurringContractStatus,
    VatType,
)
from app.domain.recurrence import validate_recurrence_fields
from app.schemas.common import ApiModel


class RecurringContractBase(ApiModel):
    label: str = Field(min_length=1, max_length=160)
    # 고객정보(공유 그룹에 저장)
    customer_name: str
    customer_phone: str
    customer_address: str
    customer_address_detail: str | None = None
    customer_visible_payment: bool = False
    notes: str | None = None
    # 스케줄
    recurrence_mode: RecurrenceMode
    day_of_month: int | None = Field(default=None, ge=1, le=31)
    interval_weeks: int | None = Field(default=None, ge=1, le=12)
    weekday: int | None = Field(default=None, ge=0, le=6)
    weekdays: list[Annotated[int, Field(ge=0, le=6)]] | None = None
    start_date: date
    end_date: date | None = None
    max_occurrences: int | None = Field(default=None, ge=1)
    # 회차 템플릿
    default_partner_id: str | None = None
    team_name: str | None = None
    team_phone: str | None = None
    service_category_id: str | None = None
    service_item_id: str | None = None
    service_name: str
    size_or_quantity: str | None = None
    service_detail: str | None = None
    special_request: str | None = None
    requested_time: str | None = None
    # per_visit: total_amount=회당 금액(월 합산=회당×방문횟수). monthly: total_amount=월 고정 금액.
    billing_mode: RecurringBillingMode = RecurringBillingMode.PER_VISIT
    partner_billing_mode: RecurringBillingMode = RecurringBillingMode.PER_VISIT
    total_amount: float | None = Field(default=None, ge=0)
    discount_amount: float = Field(default=0, ge=0)
    deposit_amount: float | None = Field(default=None, ge=0)
    balance_amount: float | None = Field(default=None, ge=0)
    vat_type: VatType | None = None
    partner_payment_amount: float | None = Field(default=None, ge=0)


class RecurringContractCreate(RecurringContractBase):
    @model_validator(mode="after")
    def _check_recurrence_fields(self) -> "RecurringContractCreate":
        validate_recurrence_fields(self.recurrence_mode, self.day_of_month, self.interval_weeks)
        return self


class RecurringContractUpdate(ApiModel):
    label: str | None = None
    customer_name: str | None = None
    customer_phone: str | None = None
    customer_address: str | None = None
    customer_address_detail: str | None = None
    customer_visible_payment: bool | None = None
    notes: str | None = None
    recurrence_mode: RecurrenceMode | None = None
    day_of_month: int | None = Field(default=None, ge=1, le=31)
    interval_weeks: int | None = Field(default=None, ge=1, le=12)
    weekday: int | None = Field(default=None, ge=0, le=6)
    weekdays: list[Annotated[int, Field(ge=0, le=6)]] | None = None
    start_date: date | None = None
    end_date: date | None = None
    max_occurrences: int | None = Field(default=None, ge=1)
    default_partner_id: str | None = None
    team_name: str | None = None
    team_phone: str | None = None
    service_category_id: str | None = None
    service_item_id: str | None = None
    service_name: str | None = None
    size_or_quantity: str | None = None
    service_detail: str | None = None
    special_request: str | None = None
    requested_time: str | None = None
    billing_mode: RecurringBillingMode | None = None
    partner_billing_mode: RecurringBillingMode | None = None
    total_amount: float | None = Field(default=None, ge=0)
    discount_amount: float | None = Field(default=None, ge=0)
    deposit_amount: float | None = Field(default=None, ge=0)
    balance_amount: float | None = Field(default=None, ge=0)
    vat_type: VatType | None = None
    partner_payment_amount: float | None = Field(default=None, ge=0)


class RecurringContractRead(RecurringContractBase):
    id: str
    order_group_id: str
    customer_token: str
    status: RecurringContractStatus
    next_due_date: date | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class RecurringContractSummaryRead(ApiModel):
    id: str
    label: str
    customer_name: str
    status: RecurringContractStatus
    schedule_text: str
    next_due_date: date | None = None

File: frontend/src/api/recurring.ts
import { apiRequest } from './client';

export type RecurrenceMode = 'monthly' | 'weekly';
export type RecurringContractStatus = 'active' | 'paused' | 'ended';

export interface RecurringContractInput {
  label: string;
  customer_name: string;
  customer_phone: string;
  customer_address: string;
  customer_address_detail?: string | null;
  customer_visible_payment?: boolean;
  notes?: string | null;
  recurrence_mode: RecurrenceMode;
  day_of_month?: number | null;
  interval_weeks?: number | null;
  weekday?: number | null;
  start_date: string;
  end_date?: string | null;
  max_occurrences?: number | null;
  default_partner_id?: string | null;
  team_name?: string | null;
  service_category_id?: string | null;
  service_item_id?: string | null;
  service_name: string;
  size_or_quantity?: string | null;
  service_detail?: string | null;
  special_request?: string | null;
  requested_time?: string | null;
  total_amount?: number | null;
  discount_amount?: number;
  deposit_amount?: number | null;
  balance_amount?: number | null;
  vat_type?: string | null;
  partner_payment_amount?: number | null;
}

export interface RecurringContractSummary {
  id: string;
  label: string;
  customer_name: string;
  status: RecurringContractStatus;
  schedule_text: string;
  next_due_date: string | null;
  pending_count: number;
  this_month_count: number;
  this_month_amount: number;
}

export interface RecurringContract extends RecurringContractInput {
  id: string;
  order_group_id: string;
  customer_token: string;
  status: RecurringContractStatus;
  next_due_date: string | null;
}

export interface PendingOccurrence {
  occurrence_id: string;
  contract_id: string;
  contract_label: string;
  customer_name: string;
  sequence_no: number;
  due_date: string;
  service_name: string;
  total_amount: number | null;
  default_partner_id: string | null;
  default_partner_name: string | null;
  is_overdue: boolean;
}

export interface ApproveItemInput {
  occurrence_id: string;
  partner_id?: string | null;
  scheduled_date?: string | null;
  total_amount?: number | null;
}

export interface ApproveOccurrencesResult {
  generated_order_ids: string[];
  skipped_occurrence_ids: string[];
}

export interface SkipItemInput {
  occurrence_id: string;
  reason?: string;
}

export interface SkipOccurrencesResult {
  skipped_occurrence_ids: string[];
}

export function listRecurringContracts(): Promise<RecurringContractSummary[]> {
  return apiRequest('/admin/recurring/contracts') as Promise<RecurringContractSummary[]>;
}

export function getRecurringContract(id: string): Promise<RecurringContract> {
  return apiRequest(`/admin/recurring/contracts/${encodeURIComponent(id)}`) as Promise<RecurringContract>;
}

export function createRecurringContract(input: RecurringContractInput): Promise<RecurringContract> {
  return apiRequest('/admin/recurring/contracts', { method: 'POST', body: input }) as Promise<RecurringContract>;
}

export function updateRecurringContract(
  id: string,
  input: Partial<RecurringContractInput>,
): Promise<RecurringContract> {
  return apiRequest(`/admin/recurring/contracts/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body: input,
  }) as Promise<RecurringContract>;
}

export function pauseRecurringContract(id: string): Promise<RecurringContract> {
  return apiRequest(`/admin/recurring/contracts/${encodeURIComponent(id)}/pause`, {
    method: 'POST',
  }) as Promise<RecurringContract>;
}

export function resumeRecurringContract(id: string): Promise<RecurringContract> {
  return apiRequest(`/admin/recurring/contracts/${encodeURIComponent(id)}/resume`, {
    method: 'POST',
  }) as Promise<RecurringContract>;
}

export function endRecurringContract(id: string): Promise<RecurringContract> {
  return apiRequest(`/admin/recurring/contracts/${encodeURIComponent(id)}/end`, {
    method: 'POST',
  }) as Promise<RecurringContract>;
}

export function deleteRecurringContract(id: string): Promise<void> {
  return apiRequest(`/admin/recurring/contracts/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  }) as Promise<void>;
}

export function syncRecurringOccurrences(): Promise<PendingOccurrence[]> {
  return apiRequest('/admin/recurring/occurrences/sync', { method: 'POST' }) as Promise<PendingOccurrence[]>;
}

export function listPendingOccurrences(): Promise<PendingOccurrence[]> {
  return apiRequest('/admin/recurring/occurrences/pending') as Promise<PendingOccurrence[]>;
}

export function approveOccurrences(items: ApproveItemInput[]): Promise<ApproveOccurrencesResult> {
  return apiRequest('/admin/recurring/occurrences/approve', {
    method: 'POST',
    body: { items },
  }) as Promise<ApproveOccurrencesResult>;
}

export function skipOccurrences(items: SkipItemInput[]): Promise<SkipOccurrencesResult> {
  return apiRequest('/admin/recurring/occurrences/skip', {
    method: 'POST',
    body: { items },
  }) as Promise<SkipOccurrencesResult>;
}
